apps/champions/lib/champions/battle.ex

```diff
@@ -10,27 +10,23 @@ defmodule Champions.Battle do
 
   @doc """
   Plays a level for a user, which means fighting its units with their selected ones.
-  Returns `:win` or `:loss` accordingly.
-
-  No tracking for level progress is done yet.
+  Returns `:win` or `:loss` accordingly, and updates the user's progress if they win..
   """
   def fight_level(user_id, level_id) do
-    user = Users.get_user(user_id)
-    level = Campaigns.get_level(level_id)
-
-    cond do
-      is_nil(user) ->
-        {:error, :user_not_found}
-
-      is_nil(level) ->
-        {:error, :level_not_found}
-
-      true ->
-        if battle(user.units, level.units) == :team_1 do
-          :win
-        else
-          :loss
+    with {:user, {:ok, user}} <- {:user, Users.get_user(user_id)},
+         {:level, {:ok, level}} <- {:level, Campaigns.get_level(level_id)} do
+      if battle(user.units, level.units) == :team_1 do
+        case Users.advance_level(user_id, level.campaign_id) do
+          # TODO: Validate level fought is the correct one before advancing - [CHoM-#214]
+          {:ok, _changes} -> :win
+          _error -> {:error, :failed_to_advance}
```

Comment:
This code is a bit messy. It's hard to notice the returning value.
We have a with, an if and a case in it. I have a few suggestions to improve its readability:
- Avoid with statement for getters, we can just pattern-match with :ok. If we want to change the msg error, we could do that in the getter function itself instead of doing it here.
- Replace the if statement with an auxiliary function. You can pattern-match the :team_1 atom to handle each return.
- For the case statement, we can do what I mentioned above for the others (pattern-match, change function returning value).

Reply:
I'm not sure I understand your suggestions here. For #1, we need this with to avoid having multiple nested ifs/cases. For #3 we're using pattern matching for the return of `Users.advance_level` like you say.

In #2 we could use a case instead of an if in order to pattern match it, but would that really make the code clearer? I think the "better" way for when you have two outcomes is an if/else rather than a case with two matches.

Reply:
This is what I could come up with to make the returned values clearer:
```
def fight_level(user_id, level_id) do
  with {:user, {:ok, user}} <- {:user, Users.get_user(user_id)},
       {:level, {:ok, level}} <- {:level, Campaigns.get_level(level_id)},
       {:campaign_progress, {:ok, %CampaignProgress{level_id: current_level_id}}} <-
         {:campaign_progress, Campaigns.get_campaign_progress(user_id, level.campaign_id)},
       {:level_valid, true} <- {:level_valid, current_level_id == level_id} do
    cond do
      battle(user.units, level.units) != :team_1 -> :loss
      advance_level(user_id, level.campaig_id) -> :win
      true -> {:error, :failed_to_advance}
    end
  else
    {:user, {:error, :not_found}} -> {:error, :user_not_found}
    {:level, {:error, :not_found}} -> {:error, :level_not_found}
    {:campaign_progress, {:error, :not_found}} -> {:error, :campaign_progress_not_found}
    {:level_valid, false} -> {:error, :level_invalid}
  end
end

defp advance_level(user_id, campaign_id) do
  case Users.advance_level(user_id, campaign_id) do
    {:ok, _changes} -> true
    _error -> false
  end
end
```
I'm not a fan of this approach. It "hides" our calls in the conditions of the `cond` and also adds an auxiliary function that doesn't really do anything. In any case, if you think this way is clearer we can go for it.